{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from importlib import reload\n",
    "import sys\n",
    "import numpy as np \n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "plt.ioff()\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "import pybedtools\n",
    "from joblib import Parallel, delayed\n",
    "import get_feature_matrix_perchr as gfm\n",
    "from tqdm import tqdm\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'get_feature_matrix_perchr' from '/home/louiscam/projects/gpcr/code/processing_regulatory_marks/IMR90_epigenome_processing/get_feature_matrix_perchr.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reload modules in case of modifications\n",
    "reload(gfm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_dir = '/home/louiscam/projects/gpcr/data/genome_data/'\n",
    "epigenome_dir = '/home/louiscam/projects/gpcr/data/regulatory_data/regulatory_data_IMR90/'\n",
    "processed_epigenome_data_dir = '/home/louiscam/projects/gpcr/save/processed_epigenome_data_dir/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = 'IMR90'\n",
    "resol_str = '250kb'\n",
    "resol = 250000\n",
    "quality = 'MAPQGE30'"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# List available genomic features and file names in a data frame\n",
    "features = ['H3K4me1','H3K4me2','H3K4me3','H3K36me3',\n",
    "            'H3K9ac','POLR2A','H3K9me3','H3K27me3',\n",
    "            'RNAseq','ATACseq']\n",
    "filenames = ['E017-H3K4me1.broadPeak',\n",
    "             'E017-H3K4me2.broadPeak',\n",
    "             'E017-H3K4me3.broadPeak',\n",
    "             'E017-H3K36me3.broadPeak',\n",
    "             'E017-H3K9ac.broadPeak',\n",
    "             'wgEncodeAwgTfbsSydhImr90Pol2IggrabUniPk.narrowPeak',\n",
    "             'E017-H3K9me3.broadPeak',\n",
    "             'E017-H3K27me3.broadPeak',\n",
    "             'GSM438363_UCSD.IMR90.mRNA-Seq.mRNA-seq_imr90_r1.bed',\n",
    "             'ENCFF243NTP.bed']\n",
    "identifier_col = [4,4,4,4,4,3,4,4,4,4]\n",
    "sources = ['RoadmapEpigenomics','RoadmapEpigenomics','RoadmapEpigenomics',\n",
    "           'RoadmapEpigenomics','RoadmapEpigenomics','ENCODE',\n",
    "           'RoadmapEpigenomics','RoadmapEpigenomics','ENCODE',\n",
    "           'ENCODE']\n",
    "accessions = ['GSE16256','GSE16256','GSE16256','GSE16256','GSE16256',\n",
    "              'GSE31477','GSE16256','GSE16256','ENCSR424FAZ','ENCSR200OML']\n",
    "df = pd.DataFrame({'feature': features,\n",
    "                   'filename': filenames,\n",
    "                   'identifier_col': identifier_col,\n",
    "                   'source': sources, \n",
    "                   'accession':accessions})\n",
    "df.to_csv(epigenome_dir + 'features.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load dataframe of feature metadata\n",
    "df_metadata = pd.read_csv(epigenome_dir+'filenames_belyaeva.csv', sep=',', header=0, index_col=None)\n",
    "features = df_metadata['name'].values\n",
    "filenames = df_metadata['filename'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "249250621\n"
     ]
    }
   ],
   "source": [
    "# Get chromosome size\n",
    "chrom = 1\n",
    "sizes_filename = genome_dir+'chrom_hg19.sizes'\n",
    "df_sizes = gfm.get_chrom_sizes(genome_dir, resol)\n",
    "chrom_size = int(df_sizes.loc[df_sizes['chr']==str(1)]['size'])\n",
    "print(chrom_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Divide the chromosome into segments of HIC_RESOLN length\n",
    "stop_pos = np.arange(resol, chrom_size + resol, resol, dtype = 'int')\n",
    "df_chrom = pd.DataFrame()\n",
    "df_chrom['chrom'] = ['chr' + str(chrom)]*len(stop_pos)\n",
    "df_chrom['start'] = stop_pos - resol\n",
    "df_chrom['stop'] = stop_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>250000</td>\n",
       "      <td>500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500000</td>\n",
       "      <td>750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>750000</td>\n",
       "      <td>1000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248250000</td>\n",
       "      <td>248500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248500000</td>\n",
       "      <td>248750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248750000</td>\n",
       "      <td>249000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249000000</td>\n",
       "      <td>249250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249250000</td>\n",
       "      <td>249500000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>998 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start       stop\n",
       "0    chr1          0     250000\n",
       "1    chr1     250000     500000\n",
       "2    chr1     500000     750000\n",
       "3    chr1     750000    1000000\n",
       "4    chr1    1000000    1250000\n",
       "..    ...        ...        ...\n",
       "993  chr1  248250000  248500000\n",
       "994  chr1  248500000  248750000\n",
       "995  chr1  248750000  249000000\n",
       "996  chr1  249000000  249250000\n",
       "997  chr1  249250000  249500000\n",
       "\n",
       "[998 rows x 3 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to bed file\n",
    "bed_chrom = pybedtools.BedTool.from_dataframe(df_chrom)\n",
    "bed_chrom_df = bed_chrom.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "df = pd.DataFrame({'feature': features, \n",
    "                   'filename': filenames})\n",
    "feature_matrix = pd.DataFrame(index = df['feature'].values, columns = bed_chrom_df['start'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 0\n",
    "df = pd.DataFrame({'feature': features, \n",
    "                   'filename': filenames})\n",
    "feature_matrix = pd.DataFrame(index = df['feature'].values, columns = bed_chrom_df['start'].values)\n",
    "f = df.loc[i, 'filename']\n",
    "feature = df.loc[i, 'feature']\n",
    "# Get bed file of the feature\n",
    "bed = pybedtools.BedTool(epigenome_dir + f).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13440</td>\n",
       "      <td>13526</td>\n",
       "      <td>SRR020294.1305618</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14671</td>\n",
       "      <td>14757</td>\n",
       "      <td>SRR020294.1136381</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14681</td>\n",
       "      <td>14767</td>\n",
       "      <td>SRR020293.507012</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16187</td>\n",
       "      <td>16271</td>\n",
       "      <td>SRR020294.663222</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16242</td>\n",
       "      <td>16271</td>\n",
       "      <td>SRR020292.104126</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12564105</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358042</td>\n",
       "      <td>59358128</td>\n",
       "      <td>SRR020293.1575673</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12564106</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358042</td>\n",
       "      <td>59358123</td>\n",
       "      <td>SRR020293.8773948</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12564107</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358086</td>\n",
       "      <td>59358172</td>\n",
       "      <td>SRR020293.5435082</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12564108</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358086</td>\n",
       "      <td>59358172</td>\n",
       "      <td>SRR020293.8178195</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12564109</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59358680</td>\n",
       "      <td>59358765</td>\n",
       "      <td>SRR020293.1922007</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12564110 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom     start       end               name score\n",
       "0         chr1     13440     13526  SRR020294.1305618     -\n",
       "1         chr1     14671     14757  SRR020294.1136381     -\n",
       "2         chr1     14681     14767   SRR020293.507012     -\n",
       "3         chr1     16187     16271   SRR020294.663222     -\n",
       "4         chr1     16242     16271   SRR020292.104126     -\n",
       "...        ...       ...       ...                ...   ...\n",
       "12564105  chrY  59358042  59358128  SRR020293.1575673     +\n",
       "12564106  chrY  59358042  59358123  SRR020293.8773948     +\n",
       "12564107  chrY  59358086  59358172  SRR020293.5435082     +\n",
       "12564108  chrY  59358086  59358172  SRR020293.8178195     +\n",
       "12564109  chrY  59358680  59358765  SRR020293.1922007     +\n",
       "\n",
       "[12564110 rows x 5 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts for this feature and this chromosome\n",
    "out = pybedtools.bedtool.BedTool.map(bed_chrom, bed, c = 4, o = 'count_distinct')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>250000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>250000</td>\n",
       "      <td>500000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>500000</td>\n",
       "      <td>750000</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>750000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1250000</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248250000</td>\n",
       "      <td>248500000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248500000</td>\n",
       "      <td>248750000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248750000</td>\n",
       "      <td>249000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249000000</td>\n",
       "      <td>249250000</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249250000</td>\n",
       "      <td>249500000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>998 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end  name\n",
       "0    chr1          0     250000     3\n",
       "1    chr1     250000     500000     0\n",
       "2    chr1     500000     750000     7\n",
       "3    chr1     750000    1000000    47\n",
       "4    chr1    1000000    1250000    41\n",
       "..    ...        ...        ...   ...\n",
       "993  chr1  248250000  248500000     0\n",
       "994  chr1  248500000  248750000     0\n",
       "995  chr1  248750000  249000000     0\n",
       "996  chr1  249000000  249250000    22\n",
       "997  chr1  249250000  249500000     0\n",
       "\n",
       "[998 rows x 4 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10/10 [01:29<00:00,  8.96s/it]\n"
     ]
    }
   ],
   "source": [
    "# Make a dataframe to store results into\n",
    "df = pd.DataFrame({'feature': features, \n",
    "                   'filename': filenames})\n",
    "feature_matrix = pd.DataFrame(index = df['feature'].values, columns = bed_chrom_df['start'].values)\n",
    "for i in tqdm(range(len(df))):\n",
    "    time.sleep(.01)\n",
    "    f = df.loc[i, 'filename']\n",
    "    feature = df.loc[i, 'feature']\n",
    "    # Get bed file of the feature\n",
    "    bed = pybedtools.BedTool(epigenome_dir + f).sort()\n",
    "    # Get counts for this feature and this chromosome\n",
    "    out = pybedtools.bedtool.BedTool.map(bed_chrom, bed, c = 4, o = 'count_distinct')\n",
    "    counts = out.to_dataframe()['name'].values\n",
    "    # Store results into matrix\n",
    "    feature_matrix.loc[feature, :] = counts\n",
    "# write feature matrix to file\n",
    "#feature_matrix.to_csv(processed_epigenome_data_dir + 'features_matrix_chr' + str(chrom) + '.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>250000</th>\n",
       "      <th>500000</th>\n",
       "      <th>750000</th>\n",
       "      <th>1000000</th>\n",
       "      <th>1250000</th>\n",
       "      <th>1500000</th>\n",
       "      <th>1750000</th>\n",
       "      <th>2000000</th>\n",
       "      <th>2250000</th>\n",
       "      <th>...</th>\n",
       "      <th>247000000</th>\n",
       "      <th>247250000</th>\n",
       "      <th>247500000</th>\n",
       "      <th>247750000</th>\n",
       "      <th>248000000</th>\n",
       "      <th>248250000</th>\n",
       "      <th>248500000</th>\n",
       "      <th>248750000</th>\n",
       "      <th>249000000</th>\n",
       "      <th>249250000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3K4me1</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>47</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>45</td>\n",
       "      <td>41</td>\n",
       "      <td>38</td>\n",
       "      <td>52</td>\n",
       "      <td>...</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>23</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>24</td>\n",
       "      <td>30</td>\n",
       "      <td>36</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me3</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>23</td>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K36me3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>48</td>\n",
       "      <td>38</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>41</td>\n",
       "      <td>40</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>54</td>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9ac</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>65</td>\n",
       "      <td>81</td>\n",
       "      <td>69</td>\n",
       "      <td>85</td>\n",
       "      <td>80</td>\n",
       "      <td>69</td>\n",
       "      <td>78</td>\n",
       "      <td>...</td>\n",
       "      <td>31</td>\n",
       "      <td>33</td>\n",
       "      <td>31</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2A</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9me3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>31</td>\n",
       "      <td>33</td>\n",
       "      <td>19</td>\n",
       "      <td>18</td>\n",
       "      <td>61</td>\n",
       "      <td>33</td>\n",
       "      <td>36</td>\n",
       "      <td>...</td>\n",
       "      <td>44</td>\n",
       "      <td>54</td>\n",
       "      <td>58</td>\n",
       "      <td>77</td>\n",
       "      <td>73</td>\n",
       "      <td>86</td>\n",
       "      <td>71</td>\n",
       "      <td>51</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K27me3</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>19</td>\n",
       "      <td>25</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNAseq</th>\n",
       "      <td>33</td>\n",
       "      <td>6</td>\n",
       "      <td>482</td>\n",
       "      <td>8092</td>\n",
       "      <td>18179</td>\n",
       "      <td>13725</td>\n",
       "      <td>8029</td>\n",
       "      <td>2028</td>\n",
       "      <td>841</td>\n",
       "      <td>1891</td>\n",
       "      <td>...</td>\n",
       "      <td>672</td>\n",
       "      <td>1405</td>\n",
       "      <td>62</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>2500</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATACseq</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66</td>\n",
       "      <td>59</td>\n",
       "      <td>65</td>\n",
       "      <td>57</td>\n",
       "      <td>43</td>\n",
       "      <td>52</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 998 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0          250000     500000     750000     1000000    1250000    \\\n",
       "H3K4me1           3          0          7         47         41         41   \n",
       "H3K4me2           0          0          3         28         20         24   \n",
       "H3K4me3           1          0          2         23         22         25   \n",
       "H3K36me3          3          0         22         48         38         30   \n",
       "H3K9ac            3          0          6         65         81         69   \n",
       "POLR2A            0          0          1          1          1          1   \n",
       "H3K9me3           4          0          5         31         33         19   \n",
       "H3K27me3          8          0          1         28         20         17   \n",
       "RNAseq           33          6        482       8092      18179      13725   \n",
       "ATACseq           5          0          0         66         59         65   \n",
       "\n",
       "          1500000    1750000    2000000    2250000      ...      247000000  \\\n",
       "H3K4me1          45         41         38         52    ...             22   \n",
       "H3K4me2          30         36         38         19    ...              8   \n",
       "H3K4me3          20         15         23         14    ...              4   \n",
       "H3K36me3         46         41         40         31    ...             54   \n",
       "H3K9ac           85         80         69         78    ...             31   \n",
       "POLR2A            1          1          1          1    ...              1   \n",
       "H3K9me3          18         61         33         36    ...             44   \n",
       "H3K27me3          5         20         19         25    ...              1   \n",
       "RNAseq         8029       2028        841       1891    ...            672   \n",
       "ATACseq          57         43         52         48    ...             19   \n",
       "\n",
       "          247250000  247500000  247750000  248000000  248250000  248500000  \\\n",
       "H3K4me1          22         23          9          3          0          0   \n",
       "H3K4me2          14         18          5          5          1          0   \n",
       "H3K4me3           7          9          1          4          1          0   \n",
       "H3K36me3         28         16          1         10          1          0   \n",
       "H3K9ac           33         31         11         11          3          5   \n",
       "POLR2A            1          0          0          0          0          0   \n",
       "H3K9me3          54         58         77         73         86         71   \n",
       "H3K27me3         18         15          0          0          0          0   \n",
       "RNAseq         1405         62         10         17          3          3   \n",
       "ATACseq          30          1          0          0          0          0   \n",
       "\n",
       "          248750000  249000000  249250000  \n",
       "H3K4me1           0         22          0  \n",
       "H3K4me2           0         16          0  \n",
       "H3K4me3           0          8          0  \n",
       "H3K36me3          1         13          0  \n",
       "H3K9ac            1         30          0  \n",
       "POLR2A            0          1          0  \n",
       "H3K9me3          51         26          0  \n",
       "H3K27me3          0         26          0  \n",
       "RNAseq            5       2500          0  \n",
       "ATACseq          32          0          0  \n",
       "\n",
       "[10 rows x 998 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process histone ChiP-seq data for all chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [32:19<00:00, 88.17s/it]\n"
     ]
    }
   ],
   "source": [
    "chr_list = np.arange(1,22+1)\n",
    "for chrom in tqdm(chr_list):\n",
    "    time.sleep(.01)\n",
    "    \n",
    "    # Get chromosome size\n",
    "    sizes_filename = genome_dir+'chrom_hg19.sizes'\n",
    "    df_sizes = pd.read_csv(sizes_filename, sep = '\\t', header = None, names=['chr','size'])\n",
    "    chrom_size = int(df_sizes.loc[df_sizes['chr']=='chr'+str(chrom)]['size'])\n",
    "    \n",
    "    # Divide the chromosome into segments of HIC_RESOLN length\n",
    "    stop_pos = np.arange(resol, chrom_size + resol, resol, dtype = 'int')\n",
    "    df_chrom = pd.DataFrame()\n",
    "    df_chrom['chrom'] = ['chr' + str(chrom)]*len(stop_pos)\n",
    "    df_chrom['start'] = stop_pos - resol\n",
    "    df_chrom['stop'] = stop_pos\n",
    "    \n",
    "    # Convert to bed file\n",
    "    bed_chrom = pybedtools.BedTool.from_dataframe(df_chrom)\n",
    "    bed_chrom_df = bed_chrom.to_dataframe()\n",
    "    \n",
    "    # Make a dataframe to store results into\n",
    "    df = pd.DataFrame({'feature': features, \n",
    "                       'filename': filenames})\n",
    "    feature_matrix = pd.DataFrame(index = df['feature'].values, columns = bed_chrom_df['start'].values)\n",
    "    for i in range(len(df)):\n",
    "        f = df.loc[i, 'filename']\n",
    "        feature = df.loc[i, 'feature']\n",
    "        # Get bed file of the feature\n",
    "        bed = pybedtools.BedTool(epigenome_dir + f).sort()\n",
    "        # Get counts for this feature and this chromosome\n",
    "        out = pybedtools.bedtool.BedTool.map(bed_chrom, bed, c = 4, o = 'count_distinct')\n",
    "        counts = out.to_dataframe()['name'].values\n",
    "        # Store results into matrix\n",
    "        feature_matrix.loc[feature, :] = counts\n",
    "        \n",
    "    # write feature matrix to file\n",
    "    feature_matrix.to_csv(processed_epigenome_data_dir + 'features_matrix_chr' + str(chrom) + '.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalize ChIP-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute mean and standard deviation across all the genome for each feature\n",
    "blacklist = {chrom: set() for chrom in np.arange(1,22+1)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect chipseq data across all chromosomes into one dataframe\n",
    "df_all = pd.DataFrame()\n",
    "chr_list =  np.arange(1,22+1)\n",
    "for chrom in chr_list:\n",
    "    df_chipseq = pd.read_csv(processed_epigenome_data_dir+'features_matrix_chr'+str(chrom)+'.csv', index_col = 0)\n",
    "    # get all blacklisted locations\n",
    "    blacklist_chr = blacklist[chrom]\n",
    "    # get a list of columns to keep\n",
    "    allcols = set(map(int,df_chipseq.columns))\n",
    "    cols2keep = allcols - blacklist_chr\n",
    "    df_chipseq_filt = df_chipseq[list(map(str,cols2keep))]\n",
    "    # Concatenate to data for previous chromosomes\n",
    "    df_all = pd.concat([df_all, df_chipseq_filt],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transform data\n",
    "df_all = np.log(df_all + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find mean and standard dev\n",
    "mean_features = np.mean(df_all, axis =1)\n",
    "std_features = np.std(df_all, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:00<00:00, 25.94it/s]\n"
     ]
    }
   ],
   "source": [
    "# Normalize ChIP-seq data\n",
    "for chrom in tqdm(chr_list):\n",
    "    time.sleep(.01)    \n",
    "    # get chipseq data\n",
    "    df_chipseq = pd.read_csv(processed_epigenome_data_dir+'features_matrix_chr'+str(chrom)+'.csv', index_col = 0)\n",
    "    # transform\n",
    "    df_chipseq = np.log(df_chipseq + 1)\n",
    "    # normalize\n",
    "    df_norm = (df_chipseq.T - mean_features)/std_features\n",
    "    # transpose back\n",
    "    df_norm = df_norm.T\n",
    "    # save\n",
    "    df_norm.to_csv(processed_epigenome_data_dir+'features_matrix_chr'+str(chrom)+'norm.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>250000</th>\n",
       "      <th>500000</th>\n",
       "      <th>750000</th>\n",
       "      <th>1000000</th>\n",
       "      <th>1250000</th>\n",
       "      <th>1500000</th>\n",
       "      <th>1750000</th>\n",
       "      <th>2000000</th>\n",
       "      <th>2250000</th>\n",
       "      <th>...</th>\n",
       "      <th>49000000</th>\n",
       "      <th>49250000</th>\n",
       "      <th>49500000</th>\n",
       "      <th>49750000</th>\n",
       "      <th>50000000</th>\n",
       "      <th>50250000</th>\n",
       "      <th>50500000</th>\n",
       "      <th>50750000</th>\n",
       "      <th>51000000</th>\n",
       "      <th>51250000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3K4me1</th>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.226778</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>-1.685095</td>\n",
       "      <td>0.443262</td>\n",
       "      <td>0.754034</td>\n",
       "      <td>0.988216</td>\n",
       "      <td>0.786294</td>\n",
       "      <td>0.860666</td>\n",
       "      <td>-1.685095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me2</th>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.112994</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>-1.655396</td>\n",
       "      <td>0.514212</td>\n",
       "      <td>0.798193</td>\n",
       "      <td>0.798193</td>\n",
       "      <td>0.923664</td>\n",
       "      <td>0.763408</td>\n",
       "      <td>-1.655396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me3</th>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-0.628214</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>-1.389419</td>\n",
       "      <td>1.139249</td>\n",
       "      <td>1.584526</td>\n",
       "      <td>2.100678</td>\n",
       "      <td>1.721978</td>\n",
       "      <td>2.053939</td>\n",
       "      <td>-1.389419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K36me3</th>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>-1.653881</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.081085</td>\n",
       "      <td>0.235184</td>\n",
       "      <td>-0.165395</td>\n",
       "      <td>-0.081085</td>\n",
       "      <td>0.374163</td>\n",
       "      <td>0.826929</td>\n",
       "      <td>0.891074</td>\n",
       "      <td>0.848955</td>\n",
       "      <td>1.070967</td>\n",
       "      <td>-1.653881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9ac</th>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>-2.475365</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.645567</td>\n",
       "      <td>-0.557825</td>\n",
       "      <td>-0.557825</td>\n",
       "      <td>-0.743654</td>\n",
       "      <td>0.676021</td>\n",
       "      <td>0.730949</td>\n",
       "      <td>1.086144</td>\n",
       "      <td>1.204543</td>\n",
       "      <td>0.906078</td>\n",
       "      <td>-2.475365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2A</th>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>-0.865347</td>\n",
       "      <td>1.155606</td>\n",
       "      <td>1.155606</td>\n",
       "      <td>1.155606</td>\n",
       "      <td>1.155606</td>\n",
       "      <td>1.155606</td>\n",
       "      <td>-0.865347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9me3</th>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>-3.301084</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.609396</td>\n",
       "      <td>-0.341489</td>\n",
       "      <td>-0.559045</td>\n",
       "      <td>-0.609396</td>\n",
       "      <td>0.134222</td>\n",
       "      <td>0.243909</td>\n",
       "      <td>0.134222</td>\n",
       "      <td>-0.303473</td>\n",
       "      <td>0.360468</td>\n",
       "      <td>-3.301084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K27me3</th>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.547482</td>\n",
       "      <td>-1.077664</td>\n",
       "      <td>0.660868</td>\n",
       "      <td>0.331789</td>\n",
       "      <td>0.711099</td>\n",
       "      <td>0.372881</td>\n",
       "      <td>0.881440</td>\n",
       "      <td>-1.547482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNAseq</th>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>-1.710399</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.463169</td>\n",
       "      <td>-0.357675</td>\n",
       "      <td>-1.202040</td>\n",
       "      <td>-1.053355</td>\n",
       "      <td>0.617851</td>\n",
       "      <td>1.139740</td>\n",
       "      <td>1.471084</td>\n",
       "      <td>1.336606</td>\n",
       "      <td>0.919753</td>\n",
       "      <td>-1.710399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 206 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0    250000    500000    750000   1000000   1250000  \\\n",
       "H3K4me1  -1.685095 -1.685095 -1.685095 -1.685095 -1.685095 -1.685095   \n",
       "H3K4me2  -1.655396 -1.655396 -1.655396 -1.655396 -1.655396 -1.655396   \n",
       "H3K4me3  -1.389419 -1.389419 -1.389419 -1.389419 -1.389419 -1.389419   \n",
       "H3K36me3 -1.653881 -1.653881 -1.653881 -1.653881 -1.653881 -1.653881   \n",
       "H3K9ac   -2.475365 -2.475365 -2.475365 -2.475365 -2.475365 -2.475365   \n",
       "POLR2A   -0.865347 -0.865347 -0.865347 -0.865347 -0.865347 -0.865347   \n",
       "H3K9me3  -3.301084 -3.301084 -3.301084 -3.301084 -3.301084 -3.301084   \n",
       "H3K27me3 -1.547482 -1.547482 -1.547482 -1.547482 -1.547482 -1.547482   \n",
       "RNAseq   -1.710399 -1.710399 -1.710399 -1.710399 -1.710399 -1.710399   \n",
       "\n",
       "           1500000   1750000   2000000   2250000    ...     49000000  \\\n",
       "H3K4me1  -1.685095 -1.685095 -1.685095 -1.685095    ...    -1.685095   \n",
       "H3K4me2  -1.655396 -1.655396 -1.655396 -1.655396    ...    -1.655396   \n",
       "H3K4me3  -1.389419 -1.389419 -1.389419 -1.389419    ...    -1.389419   \n",
       "H3K36me3 -1.653881 -1.653881 -1.653881 -1.653881    ...    -0.081085   \n",
       "H3K9ac   -2.475365 -2.475365 -2.475365 -2.475365    ...    -0.645567   \n",
       "POLR2A   -0.865347 -0.865347 -0.865347 -0.865347    ...    -0.865347   \n",
       "H3K9me3  -3.301084 -3.301084 -3.301084 -3.301084    ...    -0.609396   \n",
       "H3K27me3 -1.547482 -1.547482 -1.547482 -1.547482    ...    -1.547482   \n",
       "RNAseq   -1.710399 -1.710399 -1.710399 -1.710399    ...    -0.463169   \n",
       "\n",
       "          49250000  49500000  49750000  50000000  50250000  50500000  \\\n",
       "H3K4me1  -1.226778 -1.685095 -1.685095  0.443262  0.754034  0.988216   \n",
       "H3K4me2  -1.112994 -1.655396 -1.655396  0.514212  0.798193  0.798193   \n",
       "H3K4me3  -0.628214 -1.389419 -1.389419  1.139249  1.584526  2.100678   \n",
       "H3K36me3  0.235184 -0.165395 -0.081085  0.374163  0.826929  0.891074   \n",
       "H3K9ac   -0.557825 -0.557825 -0.743654  0.676021  0.730949  1.086144   \n",
       "POLR2A   -0.865347 -0.865347 -0.865347  1.155606  1.155606  1.155606   \n",
       "H3K9me3  -0.341489 -0.559045 -0.609396  0.134222  0.243909  0.134222   \n",
       "H3K27me3 -1.547482 -1.547482 -1.077664  0.660868  0.331789  0.711099   \n",
       "RNAseq   -0.357675 -1.202040 -1.053355  0.617851  1.139740  1.471084   \n",
       "\n",
       "          50750000  51000000  51250000  \n",
       "H3K4me1   0.786294  0.860666 -1.685095  \n",
       "H3K4me2   0.923664  0.763408 -1.655396  \n",
       "H3K4me3   1.721978  2.053939 -1.389419  \n",
       "H3K36me3  0.848955  1.070967 -1.653881  \n",
       "H3K9ac    1.204543  0.906078 -2.475365  \n",
       "POLR2A    1.155606  1.155606 -0.865347  \n",
       "H3K9me3  -0.303473  0.360468 -3.301084  \n",
       "H3K27me3  0.372881  0.881440 -1.547482  \n",
       "RNAseq    1.336606  0.919753 -1.710399  \n",
       "\n",
       "[9 rows x 206 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>250000</th>\n",
       "      <th>500000</th>\n",
       "      <th>750000</th>\n",
       "      <th>1000000</th>\n",
       "      <th>1250000</th>\n",
       "      <th>1500000</th>\n",
       "      <th>1750000</th>\n",
       "      <th>2000000</th>\n",
       "      <th>2250000</th>\n",
       "      <th>...</th>\n",
       "      <th>247000000</th>\n",
       "      <th>247250000</th>\n",
       "      <th>247500000</th>\n",
       "      <th>247750000</th>\n",
       "      <th>248000000</th>\n",
       "      <th>248250000</th>\n",
       "      <th>248500000</th>\n",
       "      <th>248750000</th>\n",
       "      <th>249000000</th>\n",
       "      <th>249250000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H3K4me1</th>\n",
       "      <td>-1.008735</td>\n",
       "      <td>-2.015639</td>\n",
       "      <td>-0.505283</td>\n",
       "      <td>0.796121</td>\n",
       "      <td>0.699133</td>\n",
       "      <td>0.699133</td>\n",
       "      <td>0.765209</td>\n",
       "      <td>0.699133</td>\n",
       "      <td>0.645307</td>\n",
       "      <td>0.868093</td>\n",
       "      <td>...</td>\n",
       "      <td>0.261757</td>\n",
       "      <td>0.261757</td>\n",
       "      <td>0.292669</td>\n",
       "      <td>-0.343208</td>\n",
       "      <td>-1.008735</td>\n",
       "      <td>-2.015639</td>\n",
       "      <td>-2.015639</td>\n",
       "      <td>-2.015639</td>\n",
       "      <td>0.261757</td>\n",
       "      <td>-2.015639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me2</th>\n",
       "      <td>-1.958933</td>\n",
       "      <td>-1.958933</td>\n",
       "      <td>-0.780395</td>\n",
       "      <td>0.903726</td>\n",
       "      <td>0.629325</td>\n",
       "      <td>0.777549</td>\n",
       "      <td>0.960422</td>\n",
       "      <td>1.110838</td>\n",
       "      <td>1.155592</td>\n",
       "      <td>0.587846</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.090994</td>\n",
       "      <td>0.343277</td>\n",
       "      <td>0.544240</td>\n",
       "      <td>-0.435694</td>\n",
       "      <td>-0.435694</td>\n",
       "      <td>-1.369664</td>\n",
       "      <td>-1.958933</td>\n",
       "      <td>-1.958933</td>\n",
       "      <td>0.449683</td>\n",
       "      <td>-1.958933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K4me3</th>\n",
       "      <td>-0.762508</td>\n",
       "      <td>-1.559915</td>\n",
       "      <td>-0.296056</td>\n",
       "      <td>2.096163</td>\n",
       "      <td>2.047202</td>\n",
       "      <td>2.188246</td>\n",
       "      <td>1.942547</td>\n",
       "      <td>1.629711</td>\n",
       "      <td>2.096163</td>\n",
       "      <td>1.555465</td>\n",
       "      <td>...</td>\n",
       "      <td>0.291605</td>\n",
       "      <td>0.832304</td>\n",
       "      <td>1.089012</td>\n",
       "      <td>-0.762508</td>\n",
       "      <td>0.291605</td>\n",
       "      <td>-0.762508</td>\n",
       "      <td>-1.559915</td>\n",
       "      <td>-1.559915</td>\n",
       "      <td>0.967804</td>\n",
       "      <td>-1.559915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K36me3</th>\n",
       "      <td>-0.846516</td>\n",
       "      <td>-1.915456</td>\n",
       "      <td>0.502253</td>\n",
       "      <td>1.085439</td>\n",
       "      <td>0.909434</td>\n",
       "      <td>0.732414</td>\n",
       "      <td>1.053306</td>\n",
       "      <td>0.966577</td>\n",
       "      <td>0.947996</td>\n",
       "      <td>0.756895</td>\n",
       "      <td>...</td>\n",
       "      <td>1.174509</td>\n",
       "      <td>0.680990</td>\n",
       "      <td>0.269171</td>\n",
       "      <td>-1.380986</td>\n",
       "      <td>-0.066493</td>\n",
       "      <td>-1.380986</td>\n",
       "      <td>-1.915456</td>\n",
       "      <td>-1.380986</td>\n",
       "      <td>0.119462</td>\n",
       "      <td>-1.915456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9ac</th>\n",
       "      <td>-2.030991</td>\n",
       "      <td>-3.572735</td>\n",
       "      <td>-1.408623</td>\n",
       "      <td>1.086720</td>\n",
       "      <td>1.328125</td>\n",
       "      <td>1.152158</td>\n",
       "      <td>1.381093</td>\n",
       "      <td>1.314479</td>\n",
       "      <td>1.152158</td>\n",
       "      <td>1.286674</td>\n",
       "      <td>...</td>\n",
       "      <td>0.281626</td>\n",
       "      <td>0.349048</td>\n",
       "      <td>0.281626</td>\n",
       "      <td>-0.809187</td>\n",
       "      <td>-0.809187</td>\n",
       "      <td>-2.030991</td>\n",
       "      <td>-1.580059</td>\n",
       "      <td>-2.801863</td>\n",
       "      <td>0.246317</td>\n",
       "      <td>-3.572735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2A</th>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>...</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>-0.914753</td>\n",
       "      <td>1.093191</td>\n",
       "      <td>-0.914753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K9me3</th>\n",
       "      <td>-4.012477</td>\n",
       "      <td>-6.973887</td>\n",
       "      <td>-3.677000</td>\n",
       "      <td>-0.596838</td>\n",
       "      <td>-0.485287</td>\n",
       "      <td>-1.461657</td>\n",
       "      <td>-1.556038</td>\n",
       "      <td>0.620154</td>\n",
       "      <td>-0.485287</td>\n",
       "      <td>-0.329699</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030477</td>\n",
       "      <td>0.399716</td>\n",
       "      <td>0.528894</td>\n",
       "      <td>1.042577</td>\n",
       "      <td>0.945711</td>\n",
       "      <td>1.243507</td>\n",
       "      <td>0.895296</td>\n",
       "      <td>0.296510</td>\n",
       "      <td>-0.909456</td>\n",
       "      <td>-6.973887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H3K27me3</th>\n",
       "      <td>-0.207024</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>-1.297760</td>\n",
       "      <td>0.641495</td>\n",
       "      <td>0.407424</td>\n",
       "      <td>0.295636</td>\n",
       "      <td>-0.501062</td>\n",
       "      <td>0.407424</td>\n",
       "      <td>0.372042</td>\n",
       "      <td>0.562305</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.297760</td>\n",
       "      <td>0.334845</td>\n",
       "      <td>0.210222</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>-1.800421</td>\n",
       "      <td>0.589674</td>\n",
       "      <td>-1.800421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNAseq</th>\n",
       "      <td>-0.580582</td>\n",
       "      <td>-1.212921</td>\n",
       "      <td>0.481148</td>\n",
       "      <td>1.608928</td>\n",
       "      <td>1.932739</td>\n",
       "      <td>1.820298</td>\n",
       "      <td>1.605801</td>\n",
       "      <td>1.055406</td>\n",
       "      <td>0.703510</td>\n",
       "      <td>1.027435</td>\n",
       "      <td>...</td>\n",
       "      <td>0.613873</td>\n",
       "      <td>0.908651</td>\n",
       "      <td>-0.333810</td>\n",
       "      <td>-1.032082</td>\n",
       "      <td>-0.835041</td>\n",
       "      <td>-1.436824</td>\n",
       "      <td>-1.436824</td>\n",
       "      <td>-1.274597</td>\n",
       "      <td>1.139086</td>\n",
       "      <td>-1.991482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATACseq</th>\n",
       "      <td>-0.649909</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>1.160163</td>\n",
       "      <td>1.077385</td>\n",
       "      <td>1.148882</td>\n",
       "      <td>1.051953</td>\n",
       "      <td>0.844721</td>\n",
       "      <td>0.984326</td>\n",
       "      <td>0.925460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.253256</td>\n",
       "      <td>0.582015</td>\n",
       "      <td>-1.474038</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>0.628915</td>\n",
       "      <td>-1.994005</td>\n",
       "      <td>-1.994005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 998 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0    250000    500000    750000   1000000   1250000  \\\n",
       "H3K4me1  -1.008735 -2.015639 -0.505283  0.796121  0.699133  0.699133   \n",
       "H3K4me2  -1.958933 -1.958933 -0.780395  0.903726  0.629325  0.777549   \n",
       "H3K4me3  -0.762508 -1.559915 -0.296056  2.096163  2.047202  2.188246   \n",
       "H3K36me3 -0.846516 -1.915456  0.502253  1.085439  0.909434  0.732414   \n",
       "H3K9ac   -2.030991 -3.572735 -1.408623  1.086720  1.328125  1.152158   \n",
       "POLR2A   -0.914753 -0.914753  1.093191  1.093191  1.093191  1.093191   \n",
       "H3K9me3  -4.012477 -6.973887 -3.677000 -0.596838 -0.485287 -1.461657   \n",
       "H3K27me3 -0.207024 -1.800421 -1.297760  0.641495  0.407424  0.295636   \n",
       "RNAseq   -0.580582 -1.212921  0.481148  1.608928  1.932739  1.820298   \n",
       "ATACseq  -0.649909 -1.994005 -1.994005  1.160163  1.077385  1.148882   \n",
       "\n",
       "           1500000   1750000   2000000   2250000    ...      247000000  \\\n",
       "H3K4me1   0.765209  0.699133  0.645307  0.868093    ...       0.261757   \n",
       "H3K4me2   0.960422  1.110838  1.155592  0.587846    ...      -0.090994   \n",
       "H3K4me3   1.942547  1.629711  2.096163  1.555465    ...       0.291605   \n",
       "H3K36me3  1.053306  0.966577  0.947996  0.756895    ...       1.174509   \n",
       "H3K9ac    1.381093  1.314479  1.152158  1.286674    ...       0.281626   \n",
       "POLR2A    1.093191  1.093191  1.093191  1.093191    ...       1.093191   \n",
       "H3K9me3  -1.556038  0.620154 -0.485287 -0.329699    ...       0.030477   \n",
       "H3K27me3 -0.501062  0.407424  0.372042  0.562305    ...      -1.297760   \n",
       "RNAseq    1.605801  1.055406  0.703510  1.027435    ...       0.613873   \n",
       "ATACseq   1.051953  0.844721  0.984326  0.925460    ...       0.253256   \n",
       "\n",
       "          247250000  247500000  247750000  248000000  248250000  248500000  \\\n",
       "H3K4me1    0.261757   0.292669  -0.343208  -1.008735  -2.015639  -2.015639   \n",
       "H3K4me2    0.343277   0.544240  -0.435694  -0.435694  -1.369664  -1.958933   \n",
       "H3K4me3    0.832304   1.089012  -0.762508   0.291605  -0.762508  -1.559915   \n",
       "H3K36me3   0.680990   0.269171  -1.380986  -0.066493  -1.380986  -1.915456   \n",
       "H3K9ac     0.349048   0.281626  -0.809187  -0.809187  -2.030991  -1.580059   \n",
       "POLR2A     1.093191  -0.914753  -0.914753  -0.914753  -0.914753  -0.914753   \n",
       "H3K9me3    0.399716   0.528894   1.042577   0.945711   1.243507   0.895296   \n",
       "H3K27me3   0.334845   0.210222  -1.800421  -1.800421  -1.800421  -1.800421   \n",
       "RNAseq     0.908651  -0.333810  -1.032082  -0.835041  -1.436824  -1.436824   \n",
       "ATACseq    0.582015  -1.474038  -1.994005  -1.994005  -1.994005  -1.994005   \n",
       "\n",
       "          248750000  249000000  249250000  \n",
       "H3K4me1   -2.015639   0.261757  -2.015639  \n",
       "H3K4me2   -1.958933   0.449683  -1.958933  \n",
       "H3K4me3   -1.559915   0.967804  -1.559915  \n",
       "H3K36me3  -1.380986   0.119462  -1.915456  \n",
       "H3K9ac    -2.801863   0.246317  -3.572735  \n",
       "POLR2A    -0.914753   1.093191  -0.914753  \n",
       "H3K9me3    0.296510  -0.909456  -6.973887  \n",
       "H3K27me3  -1.800421   0.589674  -1.800421  \n",
       "RNAseq    -1.274597   1.139086  -1.991482  \n",
       "ATACseq    0.628915  -1.994005  -1.994005  \n",
       "\n",
       "[10 rows x 998 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = pd.read_csv(processed_epigenome_data_dir+'features_matrix_chr1_norm.csv', header=0, index_col=0)\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
